File: Leetcode/树/收集树上所有苹果的最少时间.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    int ans = 0;
    bool vis[100001] = { false };
    int minTime(int n, vector<vector<int>>& edges, vector<bool>& hasApple) {
        vector<vector<int>> to(n);
        for(auto& edge:edges){
            to[edge[0]].push_back(edge[1]);
            to[edge[1]].push_back(edge[0]);
        }
        dfs(0,to,hasApple);
        return ans;
    }

    bool dfs(int from, vector<vector<int>>& to,vector<bool> hasApple){
        vis[from] = true;
        if (to[from].empty()) return hasApple[from];
        bool flag = false;
        for (auto& t : to[from]) {
            if (vis[t]) continue;
            if (dfs(t, to, hasApple)) {
                ans += 2;
                flag = true;
            }
        }
        return flag || hasApple[from];
    }
};

File: Leetcode/图论/收集树上所有苹果的最少时间.py
class Solution:
    def minTime(self, n: int, edges: List[List[int]], hasApple: List[bool]) -> int:
        graph = collections.defaultdict(set)
        for u,v in edges:
            graph[u].add(v)
            graph[v].add(u)
        
        visited = set()
        def dfs(root):
            visited.add(root)
            for nex in graph[root]:
                if nex in visited:
                    continue
                dfs(nex)
                if hasApple[nex]:
                    hasApple[root] = True
        
        dfs(0)
        res = 0
        for u,v in edges:
            if hasApple[u] and hasApple[v]:
                res += 1
        return res * 2